src/ngraph/util.cpp

```diff
@@ -270,23 +251,90 @@ ngraph::FpropCache ngraph::cache_fprop(std::shared_ptr<ngraph::Function> fprop,
         cloned_results.push_back(result);
     }
 
-    // get clone bprop parameters
-    op::ParameterVector bprop_input_params;
-    for (auto param : bprop_inputs)
+    auto get_bprop_params = [&bprop_inputs, &fprop_cache]() {
+        // get clone bprop parameters
+        op::ParameterVector bprop_input_params;
+        for (auto param : bprop_inputs)
+        {
+            bprop_input_params.push_back(
+                std::dynamic_pointer_cast<op::Parameter>(fprop_cache.node_param_map->get(param)));
+        }
+
+        // add the cached fprop nodes as inputs to bprop
+        for (auto x : fprop_cache.fprop_output_nodes)
+        {
+            bprop_input_params.push_back(
+                std::dynamic_pointer_cast<op::Parameter>(fprop_cache.node_param_map->get(x)));
+        }
+        return bprop_input_params;
+    };
+
+    fprop_cache.fprop_output_nodes.clear();
+    auto cloned_bprop_inputs = get_bprop_params();
+
+    std::unordered_map<std::shared_ptr<ngraph::Node>, std::shared_ptr<ngraph::Node>>
+        inverted_node_map;
+    for (auto kv : fprop_cache.node_param_map->get_node_map())
     {
-        bprop_input_params.push_back(
-            std::dynamic_pointer_cast<op::Parameter>(fprop_cache.node_param_map->get(param)));
+        inverted_node_map[kv.second] = kv.first;
     }
 
-    // add the cached fprop nodes as inputs to bprop
-    for (auto x : fprop_cache.fprop_output_nodes)
+    auto traverse_from_results = [](ResultVector results,
+                                    std::function<void(std::shared_ptr<Node>)> f) {
+        std::unordered_set<std::shared_ptr<Node>> instances_seen;
+        std::deque<std::shared_ptr<Node>> stack;
+
+        for (auto r : results)
```

Comment:
definitely, let's move it to `ngraph_utils.cpp` either as an extra flag to `traverse_nodes` or as a standalone function